allennlp/tests/models/biaffine_dependency_parser_test.py

```diff
@@ -53,10 +54,25 @@ def test_mst_respects_no_outgoing_root_edges_constraint(self):
         #              \/        \/
         #          B(index 1)   C(index 2)
         length = torch.LongTensor([3])
-        heads, _ = self.model._run_mst_decoding(energy, length) # pylint: disable=protected-access
+        heads, _ = decode_mst(energy[0, 0, :, :].numpy(), length.item(), has_labels=False)
         # This is the correct MST, but not desirable for dependency parsing.
-        assert heads.tolist()[0] == [-1, 0, 0]
+        assert list(heads) == [-1, 0, 0]
 
-        energy[:, :, 0, :] = 0
+        # If we run the decoding with the model, it should enforce
+        # the constraint.
         heads, _ = self.model._run_mst_decoding(energy, length) # pylint: disable=protected-access
         assert heads.tolist()[0] == [-1, 0, 1]
+
+    def test_mst_decodes_arc_labels_with_respect_to_unconstrained_scores(self):
+        energy = torch.Tensor([[0, 2, 1],
+                               [10, 0, 0.5],
+                               [9, 0.2, 0]]).view(1, 1, 3, 3).expand(1, 2, 3, 3).contiguous()
+
+        # Make the score for the root label for arcs to the root token be higher - it
+        # will be masked for the MST, but we want to make sure that the tags are with
+        # respect to the unmasked tensor.
```

Comment:
I don't really understand this comment or this test, but I'll trust you.

Reply:
Basically, there was a problem when I masked out the scores per tag for the root node. The edge weights for the MST are just the max over the labels, but for actually choosing a label, we just take the argmax from over the labels for the given arc - so for the word which is the root of the tree (i.e has the root as it's parent), all of the label scores were zero - so it was arbitrarily picking the first one. I needed to separate the choosing of the labels to use the original unmasked scores, and only use the masked scores for decoding the arcs. Can I make that comment better?

Reply:
I guess if you make it clear in the comment that this is an issue for the (ROOT, root_token) edge, that would help.  Except I thought you zeroed out all of the (*, ROOT) edges, so why is (ROOT, root_token) affected...?

Reply:
Oh wow now I have to think about this _again_

Reply:
https://github.com/allenai/allennlp/pull/1619/files#diff-9ebf9be01985b8f2d35c515d6624b0ceR519 This actually means the score matrix is [parent -> child], grrr.